### ISSUE TITLE
download logs on demand

### ISSUE THREAD

**Issue by schakrava** [rockstor/rockstor-core#762]
Provide a feature to download latest log.tgz file from the web-ui.

**Comment by MFlyer** [comment]
Hi @schakrava, can i work on this?

I can see the assignee, but think it's not active, am i wrong?

Flyer

**Comment by schakrava** [comment]
You are more than welcome to, @MFlyer. this is indeed an important feature and it's safe to assume that @gkadillak is not actively working on it.

**Comment by MFlyer** [comment]
Hi All @schakrava @phillxnet @priyaganti , here is a first preview of what planned for logs access via Rockstor Web UI [Log on demand preview](https://youtu.be/ENMv5gBtjtg)

I thought having it with just a link to download rockstor logs like a clone of actual error system, so i've decided to let user choose which logs to donwload on demand plus adding a "live log reader" - to be added modal window for loading files cat/tail

Any suggestion about layout & logs to add? :blush: Actually I've got template and view good skeleton and adding on them now it's easy, backend part is the missing one (I'm used to start always from frontend beacuse it's my "babau" and my frontend coding is a little rusty comparing to backend)

**Comment by phillxnet** [comment]
@MFlyer This is great, well done. A small suggestion to assist with understanding what's what though:
you could put (kernel) in brackets after Dmesg akin to what you have done to explain smbd/nmbd. Same could be done for the Yum entry ie (packaging) or (updates), also for Nginx ie (WebUI).
Also the addition of a general system / systemd log akin to journalctrl would be good, currently just nicer time stamps but going forward there is a greater split between these. There might also be the option to employ systemd's log filtering capabilities (extensive), but that's most likely not a first throw feature in my opinion.

Not looking as closely as your self so take all of the above with a grain of salt.

This is going to be a great addition.

**Comment by MFlyer** [comment]
Hi @phillxnet, just added your suggestions (brackets adds).

Talking about systemd-systemctl /journalctl : totally agree with you, but got some doubts for outputs

Ex. this is a `systemctl list-units -t service` just to limit items, but it still outputs a long log with long items too (fsck)

![system](https://cloud.githubusercontent.com/assets/17988902/14636318/07b35fb8-062b-11e6-969f-a55a58efddb7.jpg)


We risk to have few or too much infos, depending on how it get implemented :confused: 

We could get a <select> with services like <option>s (green bg if loaded and active, red if failing) and cast service details on select change, but probably that would overload users with datas

@schakrava looking the image, why are we running wpa_supplicant?!?

**Comment by MFlyer** [comment]
Hi all, maybe I'm getting wrong or just don't see an error, but got some problems adding new url to urls.py

added `(r'^api/systemlog/', include('storageadmin.urls.logs')),` to urls.py
added logs.py under storageadmin/urls with just one line pattern for testing import

`from django.conf.urls import patterns, url`
`from storageadmin.views import (LogsManagerView)`

`urlpatterns = patterns(`
`    '',`
`    url(r'^$', LogsManagerView.as_view(),),`
`)`

added logs_manager.py under views folder and LogsmanagerView class it's a full clone of another existing Rockstor module, so it can't be broken, but after build of py i'm getting this on browser:
`ImportError at /`
`cannot import name LogsManagerView`
`Request Method:	GET`
`Request URL:	https://rockstone/`
`Django Version:	1.6.11`
`Exception Type:	ImportError`
`Exception Value:`	
`cannot import name LogsManagerView`
`Exception Location:	/opt/build/src/rockstor/storageadmin/urls/logs.py in <module>, line 20`

Meanwhile doing my job so maybe there's an error just in front of me and i can't see it :confused:

**Comment by gkadillak** [comment]
There's a broken path at the top of your file in the imports in the logs.py file.

**Comment by MFlyer** [comment]
Don't care about my last post, I was using my class (correctly called via urls, include, etc etc) but without importing it to views/__init__ :sob: 

Ufff!!
Flyer

**Comment by MFlyer** [comment]
Hi @schakrava @phillxnet @priyaganti , let's see this (first test on "inline log reading", going to add overflow etc etc etc)

[https://youtu.be/qds2xtIN7C0](https://youtu.be/qds2xtIN7C0)

To @schakrava : found my way to hack django MVC usual structure, without `@api_view` and with a class view acting how i want (performing simple api tasks -> exec command + response plain text without care for json and letting user to have an easy readable log :hand: )

**Comment by schakrava** [comment]
The video previews look pretty good @MFlyer . Really looking forward to testing the pull request. Feel free to open it once you think it's enough for others to test.

How are you handling very large log files? Is the design to return it all in a single response? 

How are you handling rotated files like rockstor.log, rockstor.log.1, etc..?

I've looked at a couple of your commits very briefly, and see that you are using check_output, but it could block on large amounts of data. Nginx may also time out. You may want to consider using websockets for streaming log output. You would register a handler to start streaming. In the backend you can use the gevent socket server to pipe the data. See data_collector.py for example.

You are off to something great, sorry for chipping in late.

For downloading logs on the other hand, you could still use the same pattern we use currently.

**Comment by MFlyer** [comment]
Hi @schakrava , thanks for your feedbacks and suggestions!

`How are you handling very large log files? Is the design to return it all in a single response?`
`I've looked at a couple of your commits very briefly, and see that you are using check_output, but it could block on large amounts of data. Nginx may also time out. You may want to consider using websockets for streaming log output. You would register a handler to start streaming. In the backend you can use the gevent socket server to pipe the data. See data_collector.py for example`

Good suggestion, that's what i was looking for, I'll check data_collector.py ! :smiley:  - You're right when you say that actually it uses a single response - althought I'm getting good responses without timeouts or similar - It's just a "skeleton" to start with

Logs download : I was thinking about extend util.py and import it like on error handler, so we can share functionalities from the same module for other views too!

**Comment by MFlyer** [comment]
Hi @schakrava, once again thanks for the great suggestion!!
I'm playing with "my old friend" socket.io (it was 2012 when I first had a talk with the great Guillermo Rauch, asking for more documentation!) and except for an initial stop to understand how it's used on Rockstor, now it  really seems the right choice :smiley:

To all - @phillxnet @priyaganti too
Just one question
Actually I'm getting log content via subprocess.popen + stdout.read() and it's faster than check_output over nginx, but we still have to wait popen to get all content from log file

I don't think we have to worry about timeouts for logs on 500-700K size (biggest one is winbind log file and it took 13-14 secs to load), but nmbd log file is 5000K and while loading it crashed on client side (Rockstor page went totally blank on chrome)

How to procede?? I can split popen.stdout with a while loop and emit() to client line by line, but i think on a 5MB file we'll get anyway a client crash

**Comment by MFlyer** [comment]
Add related to log.nmbd size - smb log level need to be reduced to 1 after AD join otherwise nmbd will log every connection becoming enormous (on a 50+ users env on a 9to6 you get nearly 5MB log at the end of the day)

**Comment by MFlyer** [comment]
Here i am after some tests with log reading and this is what we get:

**subprocess.popen + emit of full file content** <- **NO**, browser crash with big data

**subprocess.popen + emit on an iterable**
`for content in iter(reader_output.stdout.readline,''):`
`            self.emit('logReader:logcontent', {'key': 'logReader:logcontent', 'data': content})`
Better then full content

**subprocess.popen + emit on an iterable + _spawn methods_** <- (spawning gave a great contribute :grin:) : got improvements and notebook fun starts only after first 12-13K lines read while serving ~300 emits/lines every 1-2 secs

**subprocess.popen + emit on an iterable + spawn methods + gevent.sleep(0.01)**: got a cool realtime line by line write to screen, but after ~6-7k lines browser starts blooding and top on Rockstor gave data_collector up to 4% of cpu

My final suggestion:

- we get popen + emit on an iterable obviously spawning without sleep and popen+emit get a good 300-400 emits every 1-2 secs (browser seems ok)
- add a "tail -n 200" to allow users to read last useful log lines
- on user log selection warning about cat for logs greater 500Kb (getlogsize func already done :wink: )
- add inline option to download current log (you can get a tail -n 200, but download button allow you to grab full file)

**Comment by MFlyer** [comment]
Hi all again, finally got a nice solution both for server side and client side performances:

We'll have tail -n 30, tail -n 200 (default method to read current logs) and cat -n
On backend chosen log content get always sliced in 200 lines chunks and served to client on a gevent.sleep 0,5 secs

After some testing I got a 27000 lines log in ~1 min without browser crash or notebook fun spinning.
Tried with 400 lines chunks every 0,5 sec but with bad results on client side
On a ~2000 lines log you get content rendered in a few seconds
I think this is acceptable - while avoiding user client crash
Possible integration: with logs < 2000 lines emits every 0.1 sec, will test it

Flyer

**Comment by MFlyer** [comment]
Hi all @schakrava , @phillxnet , @priyaganti , @ScarabMonkey and @maxhq too (reading your latest useful contribs hope to have your opinion too! :smiley: ), here is the video for Rockstor Log Manager I've got 5 mins ago.

What is ready?:
- Log reading + modal window beautify and progress bar (backend/frontend)
- Scripts to perform logs download (backend/frontend) already able to handle requests both from "Logs downloader" and Log Reader (actually missing download button on modal window and some js to handle it)

What to do?:
- Backend code revision (move some code from inline inside event handlers to funcs)
- Add filesize render on frontend (backend function ready)
- Add checks for 0 size or missing logs file (actually 0 size files don't raise errors, just return an empty log - see winbindd video example)

Hope you'll enjoy my work - [Rockstor Log Manager - Log Reader & Downloader](https://youtu.be/kNAYireeHww)

**Comment by MFlyer** [comment]
Hi all and @schakrava , I'd like to have a check on latest commit c94e103 : 

- Log interface has been directly added into data_collector to stay on same socket.io server
- I was thinking about more code slicing in other funcs but it isn't really essential
- Except for that, you can see there's no spawn for on_* functions and i think we don't need it (cpu, memory, disks stats, etc funcs are always running to autoupdate data, instead LogReaderNamespace on_* funcs run only on request)
- Actually rotating logs are unmanaged so we have to decide if to go with a LogManager v. 1.00 + update or wait and do it (as you prefer, not a problem for me :wink:)

**Comment by MFlyer** [comment]
> Actually rotating logs are unmanaged so we have to decide if to go with a LogManager v. 1.00 + update or wait and do it (as you prefer, not a problem for me :wink:)

To @schakrava : Finally i really don't like not having rotating logs so i'm already working on it! :smiley:  
Meanwhile we can still put out Logs Manager v 1.0 then update it (worst case: 2 weeks)

**Comment by MFlyer** [comment]
Hi all,
while working on rotating logs had fun adding some minor features
- More logs available (obviously open to suggestions for more/different logs!)
- Modal window enlarge/reduce button
- Code text resize buttons (we care about users eyes health eheh)

Here a small video, showing logs .tgz direct download from log reader too :)

https://youtu.be/vwcJvRrFirY

**Comment by maxhq** [comment]
@MFlyer This is a really great module! I like the page layout and the functions.

There is only one thing that's unclear to me: does the "Live Log Reader" update when there are new log messages coming in while it's open? I.e. is it a "tail -n 30" or a "tail -f -n 30"?

**Comment by MFlyer** [comment]
Hi @maxhq, great question : actually `tail -n 30` means what you read and not tail -f

In the early development I thought about having tail -f too, but I'm afraid about possible deadlocks

For a better understanding (@schakrava can confirm): Dashboard and widgets, like sysinfo updates, work on gevent.socketio with a nice trick (on connection some switches, one for every widget, turn to true and data is pushed to client forever till disconnection - switches auto turn to false as disconnection detected, builtin gevent.socketio method)
~~Logs reading is a little different: every action start on an on_*, where * is reading logs or building logs archive, so we would have to had a..~~

...Ok, while reading what i was writing i found my way :grin: : we can do that, just paying attention. I'll have to add a "special" log reader working with a safety switch too (one more method to stop tail -f on modal closed + switch turned to false on disconnection to prevent process running after browser closing!!!)
@phillxnet and @schakrava , what do you think about this?

**Comment by phillxnet** [comment]
@MFlyer I'm afraid I am not up enough on these web technologies to comment on the possible ramifications. I agree that a live feed would be dandy but this is already a fairly major patch so there may be some sense in moving additional functionality into another issue/pr so that we might get what has been tested so far into play. Also there are re-factoring changes ongoing in the UI code that may make additional delays impact disproportionately on merge ease. Though I haven't really looked at this element re this pr. Just noting in case.

Although @maxhq idea of a live feed has great utility, so does what you have done so far. Also I would be concerned that the testing you have done to date may be negated by feature creep this late on. However please qualify my comments here on me not being very up on the web tec side so just a general concern really.

**Comment by phillxnet** [comment]
@MFlyer Oh and what @maxhq said: this is looking really great.

**Comment by MFlyer** [comment]
Hi @maxhq , @phillxnet and @schakrava , tail -f added on d3e5df3 + 42f7421 :smiley: 

Code now need revision (optimization for redundancies) + rotating log handling

Flyer/Mirko

**Comment by MFlyer** [comment]
Hi all, finally Logs Manager coding is ended :smiley: 

To @schakrava : Going to revision code, will tell you asa PR is ready for merging/your revision :wink: 

Flyer/Mirko

**Comment by MFlyer** [comment]
Code revision finished, updates on PR #1312

**Comment by schakrava** [comment]
Fixed by #1312
